flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitReadOperator.java

```diff
@@ -0,0 +1,368 @@
+	}
+
+	@Override
+	public void processElement(StreamRecord<FileInputSplit> element) throws Exception {
+		reader.addSplit(element.getValue());
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		output.emitWatermark(mark);
+	}
+
+	@Override
+	public void dispose() {
+		super.dispose();
+
+		// first try to cancel it properly and
+		// give it some time until it finishes
+		reader.cancel();
+		try {
+			reader.join(200);
+		} catch (InterruptedException e) {
+			// we can ignore this
+		}
+
+		// if the above did not work, then interrupt the thread repeatedly
+		while (reader.isAlive()) {
+
+			StringBuilder bld = new StringBuilder();
+			StackTraceElement[] stack = reader.getStackTrace();
+			for (StackTraceElement e : stack) {
+				bld.append(e).append('\n');
+			}
+			LOG.warn("The reader is stuck in method:\n {}", bld.toString());
+
+			reader.interrupt();
+			try {
+				reader.join(50);
+			} catch (InterruptedException e) {
+				// we can ignore this
+			}
+		}
+		reader = null;
+		collector = null;
+		configuration = null;
+		format = null;
+		typeInfo = null;
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+
+		// signal that no more splits will come, wait for the reader to finish
+		// and close the collector. Further cleaning up is handled by the dispose().
+
+		if (reader != null && reader.isAlive() && reader.isRunning()) {
+			// add a dummy element to signal that no more splits will
+			// arrive and wait until the reader finishes
+			reader.addSplit(EOF);
+			reader.join();
+		}
+		collector.close();
+	}
+
+	private class SplitReader<S extends Serializable, OT> extends Thread {
+
+		private volatile boolean isRunning;
+
+		private final FileInputFormat<OT> format;
+		private final TypeSerializer<OT> serializer;
+
+		private final Object checkpointLock;
+		private final TimestampedCollector<OT> collector;
+
+		private final Object lock = new Object();
+
+		private final Queue<FileInputSplit> pendingSplits;
+
+		SplitReader(FileInputFormat<OT> format,
+					TypeSerializer<OT> serializer,
+					TimestampedCollector<OT> collector,
+					Object checkpointLock) {
+
+			this.format = checkNotNull(format, "Unspecified FileInputFormat.");
+			this.serializer = checkNotNull(serializer, "Unspecified Serialized.");
+
+			this.pendingSplits = new LinkedList<>();
+			this.collector = collector;
+			this.checkpointLock = checkpointLock;
+			this.isRunning = true;
+		}
+
+		void addSplit(FileInputSplit split) {
+			Preconditions.checkNotNull(split);
+			synchronized (lock) {
+				this.pendingSplits.add(split);
+			}
+		}
+
+		public boolean isRunning() {
+			return this.isRunning;
+		}
+
+		@Override
+		public void run() {
+			FileInputSplit split = null;
+			try {
+				while (this.isRunning) {
+
+					// get the next split to read.
+					// locking is needed because checkpointing will
+					// ask for a consistent snapshot of the list.
+					synchronized (lock) {
```

Comment:
It is true that we do not modify the queue here but, concurrently, the main thread may be adding elements. This is the reason I am taking the lock.